File: Packages/StrandAnalytics/Sources/StrandAnalytics/BreathPacer.swift
import Foundation

// BreathPacer.swift — turn a breathing pace into a deterministic inhale/exhale haptic cue list you can
// "feel" with your eyes closed, screen-off. PURE + unit-tested; the BLE layer maps each `BreathCue` onto
// the strap's actual haptic command (`AppModel.buzz(loops:)` / `send(.runHapticsPattern)`) and schedules
// the gaps with the proven `HapticClock` asyncAfter walk. No I/O here.
//
// See docs/superpowers/specs/2026-06-19-v5-haptic-biofeedback-design.md (L1 "Act (the pacer)").
//
// Felt language (identical to the shipped Breathe screen, so existing users already know it):
//   • Inhale onset → ONE light pulse  (loops: 1)
//   • Exhale onset → TWO pulses       (loops: 2)
// Each WHOOP notification buzz is a FIXED-LENGTH motor pulse — we can't vary on-time per pulse, only the
// *count* (stacked loops) and the *timing*. So the cue is encoded purely as "fire N loops at offset T".
//
// A breath cycle of `bpm` breaths/min lasts 60000/bpm ms; `inhaleFraction` splits it into inhale vs
// exhale (the calming long-exhale ratio Breathe's "Relax" preset uses is ~0.4 inhale : 0.6 exhale).
// Mirrors `HapticClock.pulses` in shape: a pure `(params) -> [Cue]` list, walkable by the BLE seam.

/// Which phase of the breath a cue marks. The on-screen orb (when the screen is on) is driven by the
/// same phase clock; screen-off, the buzz is the whole cue.
public enum BreathPhase: String, Equatable, Sendable {
    /// The start of an inhale — a single light pulse.
    case inhale
    /// A breath hold (full or empty) — no buzz by default (loops: 0).
    case hold
    /// The start of an exhale — a heavier (two-pulse) cue.
    case exhale
    /// Instruction-only beat (e.g. nostril side) — no buzz.
    case textOnly
}

/// One element of a paced-breathing haptic schedule: fire `loops` buzz loops at `offsetMs` from session
/// start, marking the onset of `phase`. The BLE layer schedules the wait then calls the proven buzz.
public struct BreathCue: Equatable, Sendable {
    /// Milliseconds from the start of the session at which to fire this cue.
    public let offsetMs: Int
    /// Which breath phase this cue marks (inhale = light, exhale = heavy, hold/textOnly = silent).
    public let phase: BreathPhase
    /// How many buzz loops to play — the felt-strength language (1 = inhale, 2 = exhale, 0 = silent).
    public let loops: Int
    /// Optional UI label for `textOnly` / nostril cues (English source; localize at the view layer).
    public let label: String?

    public init(offsetMs: Int, phase: BreathPhase, loops: Int, label: String? = nil) {
        self.offsetMs = offsetMs
        self.phase = phase
        self.loops = loops
        self.label = label
    }
}

public enum BreathPacer {

    // MARK: - Tunables (Breathe parity)

    /// Loops for an inhale onset — one light pulse, as Breathe fires today.
    public static let inhaleLoops: Int = 1
    /// Loops for an exhale onset — two pulses (heavier), as Breathe fires today.
    public static let exhaleLoops: Int = 2
    /// Default inhale fraction of the cycle — the calming long-exhale ratio (≈40:60) the "Relax" preset
    /// uses. Exhale gets the remaining 0.6.
    public static let defaultInhaleFraction: Double = 0.4
    /// Slowest / fastest paces we ever schedule (the resonance sweep band, 4.5–7 br/min). Out-of-range
    /// `bpm` is clamped so the pacer never traps or emits absurd offsets.
    public static let minBpm: Double = 3.0
    public static let maxBpm: Double = 12.0

    // MARK: - Pacer

    /// Build the haptic cue list for `cycles` full breaths at `bpm` breaths/min, splitting each cycle into
    /// inhale (`inhaleFraction`) and exhale (the remainder). One inhale cue + one exhale cue per cycle, in
    /// time order. Pure: identical inputs → identical list (the `HapticClock` precedent).
    ///
    /// `bpm` is clamped to [minBpm, maxBpm] and `inhaleFraction` to a safe (0.1…0.9) interior so each
    /// phase always carries some duration. `cycles` below 1 yields an empty schedule.
    public static func schedule(bpm: Double,
                                inhaleFraction: Double = defaultInhaleFraction,
                                cycles: Int) -> [BreathCue] {
        guard cycles >= 1 else { return [] }
        let safeBpm = min(max(bpm, minBpm), maxBpm)
        let frac = min(max(inhaleFraction, 0.1), 0.9)

        // Cycle length in ms; integer so offsets are exact and platform-identical (no float drift).
        let cycleMs = Int((60_000.0 / safeBpm).rounded())
        let inhaleMs = Int((Double(cycleMs) * frac).rounded())

        var out: [BreathCue] = []
        out.reserveCapacity(cycles * 2)
        for c in 0..<cycles {
            let base = c * cycleMs
            out.append(BreathCue(offsetMs: base, phase: .inhale, loops: inhaleLoops))
            out.append(BreathCue(offsetMs: base + inhaleMs, phase: .exhale, loops: exhaleLoops))
        }
        return out
    }

    /// Total scheduled duration (ms) of a `cycles`-breath session at `bpm` — `cycles` whole cycles. Handy
    /// for the keep-awake / session-length UI without re-deriving the cycle math.
    public static func sessionDurationMs(bpm: Double, cycles: Int) -> Int {
        guard cycles >= 1 else { return 0 }
        let safeBpm = min(max(bpm, minBpm), maxBpm)
        let cycleMs = Int((60_000.0 / safeBpm).rounded())
        return cycleMs * cycles
    }
}

File: android/app/src/main/java/com/noop/analytics/BreathPacer.kt
package com.noop.analytics

import kotlin.math.roundToInt

/*
 * BreathPacer.kt — turn a breathing pace into a deterministic inhale/exhale haptic cue list you can
 * "feel" with your eyes closed, screen-off. PURE + unit-tested; the BLE layer maps each [BreathCue] onto
 * the strap's actual haptic command ([WhoopBleClient.buzz]) and schedules the gaps. No I/O here.
 *
 * Faithful Kotlin mirror of StrandAnalytics/BreathPacer.swift — keep the cue list byte-identical to Swift
 * (cross-platform parity is the contract, pinned by matching golden-vector tests).
 * See docs/superpowers/specs/2026-06-19-v5-haptic-biofeedback-design.md (L1 "Act (the pacer)").
 *
 * Felt language (identical to the shipped Breathe screen, so existing users already know it):
 *   • Inhale onset → ONE light pulse  (loops: 1)
 *   • Exhale onset → TWO pulses       (loops: 2)
 * Each WHOOP notification buzz is a FIXED-LENGTH motor pulse — we can't vary on-time per pulse, only the
 * *count* (stacked loops) and the *timing*. So the cue is encoded purely as "fire N loops at offset T".
 *
 * A breath cycle of `bpm` breaths/min lasts 60000/bpm ms; `inhaleFraction` splits it into inhale vs
 * exhale (the calming long-exhale ratio Breathe's "Relax" preset uses is ~0.4 inhale : 0.6 exhale).
 */

/** Which phase of the breath a cue marks. The on-screen orb is driven by the same phase clock. */
enum class BreathPhase {
    /** The start of an inhale — a single light pulse. */
    INHALE,

    /** A breath hold — no buzz by default. */
    HOLD,

    /** The start of an exhale — a heavier (two-pulse) cue. */
    EXHALE,

    /** Instruction-only beat — no buzz. */
    TEXT_ONLY,
}

/**
 * One element of a paced-breathing haptic schedule: fire [loops] buzz loops at [offsetMs] from session
 * start, marking the onset of [phase]. The BLE layer schedules the wait then calls the proven buzz.
 */
data class BreathCue(
    /** Milliseconds from the start of the session at which to fire this cue. */
    val offsetMs: Int,
    /** Which breath phase this cue marks (inhale = light, exhale = heavy). */
    val phase: BreathPhase,
    /** How many buzz loops to play — the felt-strength language (1 = inhale, 2 = exhale, 0 = silent). */
    val loops: Int,
    /** Optional UI label for text-only / nostril cues. */
    val label: String? = null,
)

object BreathPacer {

    // ── Tunables (Breathe parity) ────────────────────────────────────────────
    /** Loops for an inhale onset — one light pulse, as Breathe fires today. */
    const val INHALE_LOOPS: Int = 1

    /** Loops for an exhale onset — two pulses (heavier), as Breathe fires today. */
    const val EXHALE_LOOPS: Int = 2

    /** Default inhale fraction of the cycle — the calming long-exhale ratio (≈40:60) the "Relax" preset
     *  uses. Exhale gets the remaining 0.6. */
    const val DEFAULT_INHALE_FRACTION: Double = 0.4

    /** Slowest / fastest paces we ever schedule (the resonance sweep band, 4.5–7 br/min). Out-of-range
     *  `bpm` is clamped so the pacer never traps or emits absurd offsets. */
    const val MIN_BPM: Double = 3.0
    const val MAX_BPM: Double = 12.0

    // ── Pacer ────────────────────────────────────────────────────────────────

    /**
     * Build the haptic cue list for [cycles] full breaths at [bpm] breaths/min, splitting each cycle into
     * inhale ([inhaleFraction]) and exhale (the remainder). One inhale cue + one exhale cue per cycle, in
     * time order. Pure: identical inputs → identical list (the HapticClock precedent).
     *
     * [bpm] is clamped to [MIN_BPM, MAX_BPM] and [inhaleFraction] to a safe (0.1…0.9) interior so each
     * phase always carries some duration. [cycles] below 1 yields an empty schedule.
     */
    fun schedule(bpm: Double, inhaleFraction: Double = DEFAULT_INHALE_FRACTION, cycles: Int): List<BreathCue> {
        if (cycles < 1) return emptyList()
        val safeBpm = bpm.coerceIn(MIN_BPM, MAX_BPM)
        val frac = inhaleFraction.coerceIn(0.1, 0.9)

        // Cycle length in ms; integer so offsets are exact and platform-identical (no float drift).
        val cycleMs = (60_000.0 / safeBpm).roundToInt()
        val inhaleMs = (cycleMs.toDouble() * frac).roundToInt()

        val out = ArrayList<BreathCue>(cycles * 2)
        for (c in 0 until cycles) {
            val base = c * cycleMs
            out.add(BreathCue(offsetMs = base, phase = BreathPhase.INHALE, loops = INHALE_LOOPS))
            out.add(BreathCue(offsetMs = base + inhaleMs, phase = BreathPhase.EXHALE, loops = EXHALE_LOOPS))
        }
        return out
    }

    /**
     * Total scheduled duration (ms) of a [cycles]-breath session at [bpm] — [cycles] whole cycles. Handy
     * for the keep-awake / session-length UI without re-deriving the cycle math.
     */
    fun sessionDurationMs(bpm: Double, cycles: Int): Int {
        if (cycles < 1) return 0
        val safeBpm = bpm.coerceIn(MIN_BPM, MAX_BPM)
        val cycleMs = (60_000.0 / safeBpm).roundToInt()
        return cycleMs * cycles
    }
}
